pmlc/dialect/pxa/transforms/stencil_generic.cc

```diff
@@ -0,0 +1,164 @@
+          IVLOG(1, "Requested key was: " << std::make_pair(i, curr_idx));
+          throw;
+        }
+      }
+      if (!reqs_met) {
+        continue;
+      }
+
+      // If we made it to here, this index has appropriate semantics; bind it
+      // and recurse
+      llvm::SmallVector<mlir::BlockArgument, 8> new_bound_idxs(bound_idxs);
+      new_bound_idxs.push_back(block_arg);
+      RecursiveBindIndex(new_bound_idxs, tensors);
+    }
+  }
+}
+
+void StencilGeneric::DoStenciling() {
+  // Initialization
+  // TODO: Something like?: tensors.clear();
+  {
+    auto maybe_loads_and_stores = capture();
+    if (maybe_loads_and_stores) {
+      loads_and_stores = maybe_loads_and_stores.getValue();
+    } else {
+      IVLOG(4, "Cannot Stencil: Operations fail to pattern-match.");
+      return;
+    }
+  }
+
+  llvm::SmallVector<mlir::Value, 3> tensors;
+  for (auto &load_op : loads_and_stores.loads) {
+    // TODO: It's probably fine, but I'm a little worried `getMemRef` isn't
+    // const
+    tensors.push_back(load_op.getMemRef());
+  }
+  size_t first_store_idx = tensors.size();
+  for (auto &store_op : loads_and_stores.stores) {
+    // TODO: It's probably fine, but I'm a little worried `out` isn't const
+    tensors.push_back(store_op.out());
+    // TODO: Probably should handle reduces vs. true stores in a different way
+    // if (auto reduce_op = llvm::dyn_cast_or_null<AffineReduceOp>(store_op)) {
+    //   tensors.push_back(reduce_op.out());
+    // } else if (auto true_store_op =
+    // llvm::dyn_cast_or_null<mlir::AffineStoreOp>(store_op)) {
+    //   tensors.push_back(true_store_op.getMemRef());
+    // } else {
+    //   // TODO: throw?
+    //   IVLOG(1, "Unexpected failure to load tensors from ops in stenciling");
+    //   return;
+    // }
+  }
+  std::sort(tensors.begin(), tensors.begin() + first_store_idx);
+  do { // Each load tensor permutation
+    std::sort(tensors.begin() + first_store_idx, tensors.end());
+    do { // Each store tensor permutation
+      // Add all legal permutations to legal_permutations
+      RecursiveBindIndex(llvm::SmallVector<mlir::BlockArgument, 8>(), tensors);
+    } while (std::next_permutation(tensors.begin() + first_store_idx,
+                                   tensors.end()));
+  } while (std::next_permutation(tensors.begin(),
+                                 tensors.begin() + first_store_idx));
+
+  // TODO: If we desire more complex requirements than pairwise tensor-to-index
+  // stride requirements, that function could go here
+
+  for (const auto &perm : legal_permutations) {
+    // TODO: Clean, try to get everything on the same unsigned type/width
+    auto ranges = op.getConstantRanges();
+    if (!ranges) {
+      // Can't stencil without constant ranges TODO: Check this earlier, not in
+      // the middle of some late loop!
+      IVLOG(1, "NO CONSTANT RANGES TODO");
+      break;
+    }
+    for (size_t i = 0; i < perm.indexes.size(); i++) {
```

Comment:
I dont thing this correctly goes over all combinations of tiles sizes without recursion. You might want to consider reusing the code and/or logic in autotile.

Reply:
I'll take a look, thanks

Reply:
Also, yes, agreed that this does not go over the tile sizes correctly.

Reply:
Actually, I think we should use the original logic from the stencil pass in v0:
https://github.com/plaidml/plaidml/blob/master/tile/codegen/stencil.cc#L26

I think we should basically port this to MLIR. It recursively works on N-dimensions, it handles ordered permutations, and it should be fairly generic already, so shouldn't require much tweaking for MLIR (but I might be wrong about that).

Reply:
Ok, I'll take a look at that